### PR TITLE
GH-5148 Introduce "soft fail" for corrupt ValueStore

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #5148 

Briefly describe the changes proposed in this PR:

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

### CODE REVIEW

**Comment by ebner** [comment]
Thanks for the fixes @hmottestad! I haven't had time yet to look into this, but I can contribute to the documentation if you want.

Following up on another comment from your review of my PR: adding a test for an artificially corrupted store may be difficult, I don't really know how to do this reliably. Perhaps it is possible to manipulate a NativeValue instance in the ValueStore to hold an illegal state that triggers the new soft fail functionality. Any other suggestions?

**Comment by hmottestad** [comment]
I'm actually working on it right now. I have a test that sets each byte in the values.dat file to zero, one at a time, then tries to read all the data.

File: core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeSailStoreCorruptionTest.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+			overwriteByteInFile(new File(dataDir, "values.dat"), i, 0x0);
+
+			repo.init();
+
+			List<Statement> list = getStatements();
+			assertEquals(6, list.size());
+
+		}
+
+	}
```
Comment (hmottestad) [inline core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeSailStoreCorruptionTest.java:198]: This is actually the main test that tests every single byte, except the header bytes, in the values.dat file.

File: site/content/documentation/programming/repository.md [hunk 2]
```diff
@@ -98,6 +98,9 @@ import org.eclipse.rdf4j.sail.nativerdf.NativeStore;
 Repository repo = new SailRepository(new NativeStore());
 ```
 
+In the unlikely event of corruption the system property `org.eclipse.rdf4j.sail.nativerdf.softFailOnCorruptData` can be set to `true` to
+allow the NativeStore to output CorruptValue/CorruptIRI/CorruptIRIOrBNode/CorruptLiteral objects.
+
```
Comment (hmottestad) [inline site/content/documentation/programming/repository.md:103]: Documentation.

**Comment by hmottestad** [comment]
@ebner I've added some test coverage now.

I think it might be best to also try to see what happens if the data is written to a trig or json-ld file, to test the very simple corrupt value implementations i added.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/btree/BTree.java [hunk 3]
```diff
@@ -292,6 +292,12 @@ public BTree(File dataDir, String filenamePrefix, int blockSize, int valueSize,
 			this.valueSize = buf.getInt();
 			this.rootNodeID = buf.getInt();
 
+			if (rootNodeID == 0) {
+				if (nioFile.size() > blockSize) {
```
Comment (hmottestad) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/btree/BTree.java:296]: This is a somewhat brittle approach and it's really only relevant when using the soft fail property. So maybe this should be wrapped with the property too and throw something other than an illegal state exception.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java [hunk 4]
```diff
@@ -191,12 +200,62 @@ public NativeValue getValue(int id) throws IOException {
...
+	public <T extends NativeValue & Resource> T getResource(int id) throws IOException {
+
+		NativeValue resultValue = getValue(id);
+
+		if (resultValue != null && !(resultValue instanceof Resource)) {
+			if (SOFT_FAIL_ON_CORRUPT_DATA_AND_REPAIR_INDEXES && resultValue instanceof CorruptValue) {
+				return (T) new CorruptIRIOrBNode(revision, id, ((CorruptValue) resultValue).getData());
+			}
+			logger.warn(
+					"Possible corrupt data consider setting the system property org.eclipse.rdf4j.sail.nativerdf.softFailOnCorruptDataAndRepairIndexes to true");
```
Comment (aschwarte10) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java:230]: this message sounds a bit strange (though it is clear what it means. Maybe you can phrase it in two sentences, same in line 255 and line 629

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java [hunk 5]
```diff
@@ -534,49 +601,76 @@ private NativeValue data2value(int id, byte[] data) throws IOException {
 		case LITERAL_VALUE:
 			return data2literal(id, data);
 		default:
-			throw new IllegalArgumentException("Invalid type " + data[0] + " for value with id " + id);
+			if (SOFT_FAIL_ON_CORRUPT_DATA_AND_REPAIR_INDEXES) {
+				logger.error("Soft fail on corrupt data: Invalid type {} for value with id {}", data[0], id);
+				return new CorruptUnknownValue(revision, id, data);
+			}
+			throw new SailException("Invalid type " + data[0] + " for value with id " + id
```
Comment (aschwarte10) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java:608]: Maybe also here, two sentences.

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java [hunk 6]
```diff
@@ -197,29 +203,50 @@ public byte[] getData(long offset) throws IOException {
 				(data[2] << 8) & 0x0000ff00 |
 				(data[3]) & 0x000000ff;
 
-		// We have either managed to read enough data and can return the required subset of the data, or we have read
-		// too little so we need to execute another read to get the correct data.
-		if (dataLength <= data.length - 4) {
+		// If the data length is larger than 750MB, we are likely reading the wrong data. Probably data corruption.
```
Comment (aschwarte10) [inline core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java:206]: how did you come up with 750MB? Is this a special size?

**Review by aschwarte10 — COMMENTED** [review]
I am too distant from the native store implementation and its index structures to give a proper review. From what I can tell with going through the PR the logic seems fine: I can follow the idea and the new setting is opt-in. 
Please see inline for minor (optional) comments on some messages

File: core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java (hunk 6 above)
Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java:206]: It was what I experienced when testing.

I wouldn't think that a single literal would realistically be 750 MB. But I'll double check that the logic only applies if the system property is activated.

Comment (hmottestad) [reply core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java]: I added some more context to the comment

**Pull request merged**